### ISSUE TITLE
[Question] Set action on double tapping a key

### ISSUE THREAD

**Issue by rjshrjndrn** [rvaiya/keyd#219]
First of all, thank you for this wonderful project. It makes people living in keyboard, like myself, life heaven.

I've a question

Is it possible to define action on pressing a key twice?

For example, 

If I press `capslock` twice it should emulate a layer, so that I can do `h,j,k,l` for navigation, and comes out of the layer, once I press capslock again.

The idea is to have `capslock` press and hold as control for all key combos and use `h,j,k,l` for navigation.

This is my current setup

```
# To get the proper key names
# `keyd -m`
[ids]

*

[main]

# Maps capslock to escape when pressed and control when held.
# capslock = overload(control, esc)
capslock = overload(nav, esc)
leftmeta = overload(apps, leftmeta)

# Remaps the escape key to capslock
esc = capslock

[nav:C]
j = down
l = right
h = left
k = up

# Media
n = nextsong
p = previoussong
space = playpause

# VIM shortcuts
' = macro(:q enter)
; = macro(:w enter)
/ = macro(:q! enter)

# Workaround for gnome wayland
[apps:M]
m = macro(leftmeta k i t t y enter)

# tmux shortcuts
c = macro(leftcontrol+leftshift+c)
v = macro(leftcontrol+leftshift+v)
g = macro(set space -o space vi enter)
```

**Comment by slakkenhuis** [comment]
> Is it possible to define action on pressing a key twice?

Yeah --- if the first keypress takes you into one layer, then pressing it again while in that layer would correspond to having pressed the key twice:

```
[main]
capslock = oneshot(nav)

[nav]
capslock = <some action>
```

> If I press capslock twice it should emulate a layer, so that I can do h,j,k,l for navigation, and comes out of the layer, once I press capslock again.
> The idea is to have capslock press and hold as control for all key combos and use h,j,k,l for navigation.

If I understand you correctly, you want to have capslock act as the control modifier, but additionally emulate the navigation keys if you tap and then hold it a second time? If so:

```
[main]
capslock = oneshot(control)

[control:C]
capslock = oneshot(nav)

[nav:C]
j = down
l = right
h = left
k = up
```

This would mean that you lose your overloaded escape key on capslock. This is unavoidable: how would keyd know that a single tap of the capslock was just a stepping stone towards a second press, instead of escape? If you want to have both, you'll have to resort to timeouts.

**Comment by rjshrjndrn** [comment]
> > If I press capslock twice it should emulate a layer, so that I can do h,j,k,l for navigation, and comes out of the layer, once I press capslock again.
> > The idea is to have capslock press and hold as control for all key combos and use h,j,k,l for navigation.
The idea was to have capslock as sticky key, means double tap capslock, and then you're free to press any other key without holding the capslock

> This would mean that you lose your overloaded escape key on capslock. This is unavoidable: how would keyd know that a single tap of the capslock was just a stepping stone towards a second press, instead of escape? If you want to have both, you'll have to resort to timeouts.
Yes, I'll have to look for time out. I'm thinking about giving some small timeout, because double tap will be fast enough.

**Comment by slakkenhuis** [comment]
Ah. Then I suppose you'll want something like:

```
[main]
capslock = oneshot(control)

[control:C]
capslock = toggle(nav)

[nav]
capslock = toggle(main)
```

You see why? (If you don't like the control-oneshot side effect, you can achieve something similar with `capslock = overload(control, oneshot(capslocklayer))`)

> I'm thinking about giving some small timeout, because double tap will be fast enough.

I don't think timeouts *between taps* are supported, only timeouts on key *hold*. It's interesting to entertain the idea of this as a feature, though it's probably wise to be conservative about it.

**Comment by w-jablonski** [comment]
Hi and thanks for this great tool!
> I am open to persuasion on the point if someone is willing to make a case for it.
I would really like to have something like this implemented, i.e. touble tap => activate layer. Use case: like already mentioned, quickly turn a part of your keyboard into a text/terminal navigation tool. Current solution nessecitates that a single tap of that key already has an effect. This is prone to accidental mis-use - if you're using that modifier a lot, you will sometimes tap it and abandon, for example when changing your mind. Now a quick double-tap is a much more intentional action, giving a peace of mind. Thanks!

**Comment by w-jablonski** [comment]
If anyone ever needs it, I have resolved this using the currently available options, something like below:

`[global]
oneshot_timeout = 200
[main]
leftalt = overload(abc, oneshot(dummy)) # using overload, leftalt continuously serves as a modifier giving the abc layer
[dummy]
leftalt = overload(abc, toggle(abc))
[abc]
leftalt = toggle(abc)
`

Thanks! 

(Btw I did experience strange one-time fails of the keys bound to " and ' -- I have now replaced ' with apostrophe. If the " keeps playing up on newest git, I will submit a bug. So far so good.) 

Thanks again!
